src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs

```diff
@@ -0,0 +1,42 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DocAsCode.Common
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Collections.Concurrent;
+
+    public class SharedObjectManager<TState, TOperand>
+    {
+        private readonly ConcurrentDictionary<TState, SharedObjectNode<TState, TOperand>> _states;
+        private readonly IEqualityComparer<TState> _stateComparer;
+        private readonly Func<TState, TOperand, TState> _modifier;
+        private readonly Func<TState, SharedObjectNode<TState, TOperand>> _creator;
```

Comment:
avoid new func per call

---
In reply to: [120027775](https://github.com/dotnet/docfx/pull/1723#discussion_r120027775) [](ancestors = 120027775)